### DISCUSSION TITLE
Where is the source for iperf3-darwin?

### DISCUSSION THREAD

**Discussion by nirs** [esnet/iperf#1802] [General]
On macOS there is a customize iperf3 program named iperf3-darwin:

```
IPERF3-DARWIN(1)            General Commands Manual            IPERF3-DARWIN(1)

NAME
     iperf3-darwin – perform network throughput tests

SYNOPSIS
     iperf3-darwin -s [options]
     iperf3-darwin -c server [options]
     iperf3-darwin -h
     iperf3-darwin -v

DESCRIPTION
     The iperf3-darwin is a tool for performing network throughput measurements
     based on the ESnet iperf3 project that has been modified to use
     functionalities specific to Darwin.

     For a full list of options see:

           iperf3-darwin -h

AUTHORS
     A list of the contributors to iperf3 can be found within the documentation
     located at https://software.es.net/iperf/dev.html#authors.

SEE ALSO
     https://software.es.net/iperf

Darwin                           July 22, 2022                           Darwin
```

We failed to find it in https://opensource.apple.com/releases/ or https://github.com/apple-open-source/macos/

**Comment by bmah888** [comment]
Fascinating. I had no idea this even existed.

**Comment by aidan-gibson** [comment]
It's just a macOS bundled version more or less equiv to conventional iperf3? Just an older version. (my iperf3 is from homebrew)
<img width="862" alt="image" src="https://github.com/user-attachments/assets/b3acf8e8-b263-46c9-b16c-b908a0894920">

**Reply by bmah888** [reply]
Actually `iperf3-darwin` has a number of macOS specific features added on top of the version we released from ESnet. If you compare the `--help` output between the two programs you'll see quite a lot of new functionality added by Apple. I think it's unlikely the source for these changes would be released by Apple, and even if so, they might not be portable to other supported iperf3 platforms. (Note that iperf3 is BSD-licensed, so there is no obligation for Apple to release the source code for their derivative work.)

**Reply by nirs** [reply]
Indeed iperf3 has many Apple specific options:

```
  --apple-quic              use QUIC rather than TCP
  --apple-quic-verbose      enable QUIC verbose logging
  --apple-quic-multiplex    when using -P, use multiple streams instead
  --apple-tclass class      set the SO_TRAFFIC_CLASS socket option to the class
  --apple-rcvanyif          set the SO_RECV_ANYIF socket option
  --apple-boundif ifname    set the IP[V6]_BOUND_IF socket option for the interface name
  --apple-intcoproc         set the SO_INTCOPROC_ALLOW socket option
  --apple-delegate proc     set the SO_DELEGATED socket option to the process
  --apple-multiscope ifname set the IP_MULTICAST_IFINDEX socket option for the interface name
  --apple-multicastif addr  set the IPV6_MULTICAST_IF socket option with the address
  --apple-msgcnt num        number of concurrent messages
  --apple-noflowadvisory n  boolean to disable flow control advisory for UDP
  --apple-nwconnection      use the nw_connection api
  --apple-nwprotocol        use the nw_protocol api
  --apple-multipathsvc      use MPTCP, with the specified service-type
  --apple-control-host <hostname>   connect to this host for the control tcp connection only
  --apple-control-port <port>       connect to this port for the control tcp connection only
  --apple-tls               use TLS
  --apple-checksum          compute and report checksum
  --apple-busytime <X ms>   only read/write on the client-side for X milliseconds
  --apple-sleeptime <X ms>  after each busy-time (see above), sleep for X milliseconds
  --apple-ignore-order      do not emit an error if an out of order packet is received
  --apple-tempdir <path>    path of existing directory to store temporary files
  --apple-proxy <json>      proxy parameter as JSON data
  --apple-proxy @<path>     path to file with proxy parameters JSON data
  --apple-l4s               enable L4S on the connection, currently supported only for QUIC
  --apple-sleepatexit <n>   sleep at exit either indefinitely by default or optionally for <n> seconds
```

But the reason we are interested in it is that on macos-15 github runners, we see failures to access a Ubuntu 24.04 linux vm running iperf3 with iperf3 from brew. Using iperf3-darwin we don't see these issues. We don't see any issue when running the same test on local macOS 14.7.1 and 15.1.1.

This is probably some issue in the macos-15 image in the github runners, but the fact that iperf3-darawin works and the one from brew does not means something is different in iperf3-darwin.

See https://github.com/lima-vm/socket_vmnet/issues/85

**Reply by nirs** [reply]
I opened https://feedbackassistant.apple.com/feedback/16032747 (it is probably not accessible), requesting to share iperf3-darwin source.

**Reply by aidan-gibson** [reply]
Ah nice, good to know! Thanks for opening that.
